### PR TITLE
Fix #7188: AI instance crash when reloading AI in a server.

### LINKED ISSUE #7188: Both server and client crash, AI related
## Version of OpenTTD
2019-02-06-master-gbfdad9ad1b
## Expected result
No crash
## Actual result
Crash
## Steps to reproduce
1 - on server, start multiplayer non-dedicated server game with 14 AIs set up, with 250000 #opcodes, build speed set to very fast, and a start_date of 0 on every AI.
2 - on client, join server as Spectator while not all companies have started yet.
3 - on server, open AI Debug Window, click Reload on one of the AIs. If it doesn't crash right away, try Reload on another AI, until one of them crashes.
4 - on client, a timeout of 20 seconds is displayed, then it crashes after that, I suppose when exiting to main menu, not sure.

[crash.log](https://github.com/OpenTTD/OpenTTD/files/2838675/crash.log)
[crash.log](https://github.com/OpenTTD/OpenTTD/files/2838736/crash.log)

Left = server, Right = client

### PR DESCRIPTION
**Pull request by SamuXarick**

### CODE REVIEW

File: src/ai/ai_instance.cpp [hunk 1]
```diff
@@ -258,7 +258,7 @@ void CcAI(const CommandCost &result, TileIndex tile, uint32 p1, uint32 p2)
 	 * when the company does not exist anymore.
 	 */
 	const Company *c = Company::GetIfValid(_current_company);
-	if (c == NULL || c->ai_instance == NULL) return;
+	if (c == NULL || c->ai_instance == NULL || !c->ai_instance->IsWaiting()) return;
```
Comment (PeterN) [inline src/ai/ai_instance.cpp]: This doesn't add any checks to make sure it's the same ai_instance as that which called the command in the first place.

File: src/network/network_client.cpp [hunk 2]
```diff
@@ -1171,9 +1171,9 @@ void ClientNetworkGameSocketHandler::CheckConnection()
 	/* 20 seconds are (way) more than 4 game days after which
 	 * the server will forcefully disconnect you. */
 	if (lag > 20) {
+		ClientNetworkEmergencySave();
```
Comment (glx22) [inline src/network/network_client.cpp]: as CloseConnection() also save and show the error message, maybe just remove them here

**Comment by TrueBrain** [comment]
Can you check if #7298 fixes #7188 too? So instead of this PR, that #7298 is also sufficient? Tnx!

**Comment by SamuXarick** [comment]
> Can you check if #7298 fixes #7188 too? So instead of this PR, that #7298 is also sufficient? Tnx!

It fixes the client crash. The server crash still happens.

**Comment by glx22** [comment]
I have https://github.com/glx22/OpenTTD/commits/ccai_check that may be a solution, but not really tested.

**Comment by LordAro** [comment]
I'm going to close this, as there's been no activity on it for a few months. Between Peter's comments & glx's branch, I agree that the other branch looks like a better (and more general) solution.

@glx22 would you mind PRing that?

**Pull request closed without merging**